combinetf2/physicsmodels.py

```diff
@@ -0,0 +1,399 @@
+
+
+class Project(PhysicsModel):
+    """
+    A class to project a histogram to lower dimensions.
+    Optionally the histogram can be normalized.
+    The normalization is done to the integral of all processes or data.
+
+    Parameters
+    ----------
+    channel_name : str
+        Name of the channel. Required.
+    axes_names : list of str, optional
+        Names of the axes to keep. If empty, the histogram will be projected to a single bin.
+    """
+
+    def __init__(self, indata, channel, *axes_names, normalize=False):
+        info = indata.channel_info[channel]
+
+        self.start = info["start"]
+        self.stop = info["stop"]
+        self.normalize = normalize
+        self.name = "normalize" if normalize else "project"
+
+        channel_axes = info["axes"]
+
+        hist_axes = [axis for axis in channel_axes if axis.name in axes_names]
+
+        if len(hist_axes) != len(axes_names):
+            raise ValueError(
+                f"Hist axes {[h.name for h in hist_axes]} != {axes_names} not found"
+            )
+
+        exp_axes = channel_axes.copy()
+
+        # for per process histograms
+        exp_axes.append(indata.axis_procs)
+        extra_axes = [indata.axis_procs]
+
+        self.exp_shape = tuple([len(a) for a in exp_axes])
+
+        channel_axes_names = [axis.name for axis in channel_axes]
+        extra_axes_names = [axis.name for axis in extra_axes]
+
+        axis_idxs = [channel_axes_names.index(axis) for axis in axes_names]
+
+        self.proj_idxs = [i for i in range(len(channel_axes)) if i not in axis_idxs]
+
+        post_proj_axes_names = [
+            axis for axis in channel_axes_names if axis in axes_names
+        ] + extra_axes_names
+
+        self.transpose_idxs = [
+            post_proj_axes_names.index(axis) for axis in axes_names
+        ] + [post_proj_axes_names.index(axis) for axis in extra_axes_names]
+
+        self.has_data = not info["masked"]
+
+        self.channel_info = {
+            channel: {
+                "axes": hist_axes,
+                "start": None,
+                "stop": None,
+            }
+        }
+
+        self.identifiers = [
+            self.name,
+            "_".join(axes_names) if len(axes_names) else "yield",
+        ]
+
+    def make_fun(self, fun_flat, inclusive=True):
+
+        if inclusive:
+            out_shape = self.exp_shape[:-1]
+            transpose_idxs = self.transpose_idxs[:-1]
+        else:
+            out_shape = self.exp_shape
+            transpose_idxs = self.transpose_idxs
+
+        def fun():
+            exp = fun_flat()[self.start : self.stop]
+            if self.normalize:
+                norm = tf.reduce_sum(exp)
+                exp /= norm
+            exp = tf.reshape(exp, out_shape)
+            exp = tf.reduce_sum(exp, axis=self.proj_idxs)
+            exp = tf.transpose(exp, perm=transpose_idxs)
+
+            return exp
+
+        return fun
+
+    def get_data(self, data, cov=None):
```

Comment:
Should the cov parameter be part of the api in general?  (ie should the base class have this as well?)

Reply:
You mean a member of the PhysicsModel? I think it should be treated in the same way as the data, i.e. a member of the fitter or indata class that get's passed in the get_data function.

Reply:
I meant why doesn't the base class version of get_data also have a "cov" parameter?

Reply:
Done